public/javascripts/jquery.livequery.min.js

```diff
@@ -1,9 +1,8 @@
-/* Copyright (c) 2010 Brandon Aaron (http://brandonaaron.net)
```

Comment:
Since we're using jQuery 1.9, can we switch to `.on` and outright remove this plugin instead of updating it?

There's another instance of livequery use we can replace here:

https://github.com/otwcode/otwarchive/blob/58ac2bec9cf942e60911a54ff96f951d0fe1de76/public/javascripts/application.js#L53-L56

I don't think there's an easy way to test scrolling functionality in Cucumber, we'll just manually test that.

Reply:
I would love to help remove this plugin! Unfortunately I don't _think_ it's easy to replace that particular usage of livequery with `.on`.

My understanding is that [`on`](https://api.jquery.com/on/) only handles attaching event handlers to a set of elements; that behavior was duplicated in the older version of the plugin but [doesn't exist in the newer version](https://github.com/hazzik/livequery/blob/master/README.md).

The new version of the plugin still exists to support the syntax currently being used for autocomplete, which allows specifying callbacks for when a matching element is added to or removed from the DOM. From what I can tell, there is a Mutation Observer API that could be used to monitor those changes instead, [as long as you don't need to support IE < 11](https://caniuse.com/#feat=mutationobserver), but [it's not using events](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) so I'm pretty sure it would not be compatible with `.on()`. [Here's a short blog post on how to use MutationObserver.](https://gabrieleromanato.name/jquery-detecting-new-elements-with-the-mutationobserver-object)

If you'd like, I would be happy to rewrite the autocomplete initializer to use MutationObserver or some other methodology so that the plugin can be removed fully?

Reply:
`.on` could handle future elements if you use delegated event handlers. We do that with modals for example:

https://github.com/otwcode/otwarchive/blob/63adfaf4641a6fc9c2e24911d86a563fe8867508/public/javascripts/ao3modal.js#L312-L313

Reply:
@redsummernight sorry, just saw your reply. Yes, `.on()` allows you to attach i.e. click events to elements that are added to the DOM later.

What livequery allows is attaching a callback to the "event" of a new element being added or removed (which can't be handled with `.on()` because it's not a standard javascript event.) There are alternative implementations of this in modern browsers, as I've described above.

In the case of autocomplete, livequery is still being used to say that whenever an autocomplete input is added to the DOM, it should be initialized. I could replace that with MutationObserver if the browser support is sufficient.

However, I'm pretty sure I can't replace it with `.on()` – is there a specific event that you think I should be capturing that way?

Reply:
Thanks for looking into this. You're right that `.on()` needs specific events, and moving away from the plugin is more complicated than I thought. For now, let's stick with updating livequery.

Reply:
sounds good, thanks!